man1/soya.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SOYA 1"
.TH SOYA 1 "2011-03-11" "perl v5.10.1" "SOI Platform Applications Container"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
.Vb 1
\&  soya \- SOI Platform Applications Container
.Ve
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&  soya
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
.Vb 3
\&  Soya se puede usar  para  iniciar  cualquier proceso via el comando screen,
\&  es un metascript que indicando los comandos en un archivo de configuracion,
\&  facilita la ejecucion de los mismos.
\&
\&  Utiliza un directorio de configuracion, donde se almacenan los siguientes
\&  archivos:
\&    * soya.sh
\&      Aplicacion principal
\&
\&    * soya.conf
\&      Archivo de configuracion general
.Ve
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.Vb 5
\&  Para activar una nueva aplicacion es necesario hacer lo siguiente:
\&    1.  Crear una liga con el nombre de la aplicacion desde el directorio
\&        $HOME/soya/setup con el parametro start y stop, ej.
\&          ln \-sf $HOME/soya/soya.sh $HOME/soya/setup/app.start 
\&          ln \-sf $HOME/soya/soya.sh $HOME/soya/setup/app.stop 
\& 
\&    2.  Crear un archivo de configuracion, donde se incluyan algunos de los
\&        parametros necesarios para la aplicacion, ej:
\&          # app.conf \-\- put here a short description.
\&          # \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
\&          APDEBUG=off
\&          APTYPE=CT
\&          APCOMMAND="sh ${HOME}/tomcat/bin/catalina.sh run"
.Ve
.SH "AUTHOR"
.IX Header "AUTHOR"
.Vb 2
\&  soya v2.06 
\&  (c) 2009, 2010, 2011 Nextel de Mexico
\& 
\&  Developed by 
\&  Andres Aquino <aquino@hp.com>
.Ve
.SH "BUGS"
.IX Header "BUGS"
.Vb 1
\& Muchos!
\&
\& Se necesitan traductores... !!!
.Ve


HISTORY
  ___ ___ _  _ __ _ 
 (_-</ _ \ || / _` |
 /__/\___/\_, \__,_|
          |__/      
SOI Platform Applications Container
=-=

INFORMATION
	Usar las banderas correspondientes para indicar que tipo de issue fue implementado
		 DEFECT         : Report a Software Defect
		 ENHANCEMENT    : Request for a Enhancement
		 TASK           : Work item that change code or documentation
		 REVIEW         : Request for a source code review
		 OTHER          : Some other kind of issue

ROADMAP
   {branch}
      {type}: {comment}
      {documentation}
   v2.03-201011.09
      TASK: actualizar documentacion en man y para usuarios


   v2.03-201011.08
      ENHANCEMENT: port de HPUX hacia Linux
      Algunas funciones no corresponden con los mismos datos entre HPUX y Linux, por lo tanto hay
      que migrarlas y hacer las correspondientes validaciones


# vim: set ts=3 et sw=3 sts=3 si ai tw=100 lbr:


soya.sh
#!/bin/sh
# vim: set ts=2 sw=2 sts=2 si ai et: 

# soya.sh - SOI Platform Applications Container
# =-=
#
# Developer
# Andres Aquino <aquino@hp.com>
# 


#
# default enviroment
APLINK=`basename ${0%.*}`
APLEVL="DEBUG"
APTYPE="AP"
VIEWLOG=false
VIEWMLOG=false

# applications enviroment
. ${HOME}/.soyarc

# load user functions
. ${APPATH}/libutils.sh
set_environment

# soya's environment
SCRPRCS=0
VERSION="`cat ${APPATH}/VERSION | sed -e 's/-rev/ Rev./g'`"
RELEASE=`openssl dgst -md5 ${APPATH}/${APNAME}.sh | rev | cut -c-4 | rev`
SCRPRCS=`echo ${APLINK} | sed -e "s/[a-zA-Z\.-]/0/g;s/.*\([0-9][0-9]\)$/\1/g"`
APTYPE="AP"
[ ${APLINK} != ${APNAME} ] && APTYPE=`grep -i aptype ${APLINK}.conf | sed -e "s/.*=//g" 2> /dev/null`

# send version
log_action "DEBUG" "You're using ${APNAME} ${VERSION} release ${RELEASE}"

# virtual terminal name
${APDEBUG} && APDEBUG="on" || APDEBUG="off"
APACTION=`basename ${0#*.} | tr "[:lower:]" "[:upper:]"`
SCRNAME=`echo ${APHOST} | rev | cut -c 1-4 | rev`
SCRNAME=`echo ${SCRNAME}${APTYPE}${SCRPRCS} | tr "[:lower:]" "[:upper:]"`

# workaround: please check this variable in *.conf
scrName=${SCRNAME}

# application's environment
[ ${APLINK} != ${APNAME} ] && . ${APPATH}/setup/${APLINK}.conf

# check deprecated environment
[ ${#apType}  -ne 0 ] && APTYPE=${apType} && log_action "DEBUG" "DEPRECATED: Please change to uppercase apType variable"
[ ${#apCommand} -ne 0 ] && APCOMMAND=${apCommand} && log_action "DEBUG" "DEPRECATED: Please change to uppercase apCommand variable"

#
set_proc "${SCRNAME}"
get_process_id "${SCRNAME}"


##
executeCmd () {
  local STRCOMMAND="${1}"

  # eliminar referencias nulas del screen
  ${SCREEN} -wipe > /dev/null 2&>1

  # verificar si ya existe una terminal
  process_running
  [ ! -s ${APLOGT}.pid ] && report_status "i" "Hey, VirtualTerminal ${SCRNAME} doesnt exists..."
  [ ! -s ${APLOGT}.pid ] && log_action "DEBUG" "Hey, VirtualTerminal ${SCRNAME} doesnt exists..."
  [ ! -s ${APLOGT}.pid ] && exit 1

  # ejecutar el comando
  log_action "DEBUG" "Executing ${STRCOMMAND} in screen ${SCRNAME}"
  ${SCREEN} -x ${SCRNAME} -p 0 -X stuff "$(printf '%b' "${STRCOMMAND}\015")"
  wait_for "Executing command ${STRCOMMAND} on ${SCRNAME} " 2

  # reportar el estado de la ejecucion
  log_action "DEBUG" "${STRCOMMAND} on ${SCRNAME} cooked, go home baby! "
  report_status "*" "Command ${STRCOMMAND} on ${SCRNAME} executed, go home baby!"
 
}

##
get_tree_of_applications () {
  process_running
  
  if [ -s ${APLOGT}.pid ]
  then
    cat ${APLOGT}.pid | sort -n | head -n1 > ${APLOGT}.proc
    while true
    do
      APID=`tail -n1 ${APLOGT}.proc`
      APPID=`awk -v pid=${APID} -v pos=${PSPOS} '{if ($(4+pos) ~ pid){print $(3+pos)}}' ${APLOGT}.allps 2> /dev/null | sed -e "s/ *//g"`
      [ ${#APPID} -eq 0 ] && break
      echo ${APPID} >> ${APLOGT}.proc
    done
    sort -n ${APLOGT}.proc > ${APLOGT}.list
  else
    log_action "ERR" "${SCRNAME} process doesn't exist!"
  fi

}


# START
if [ ${APACTION} = "START" ]
then

  # eliminar referencias nulas del screen
  ${SCREEN} -wipe > /dev/null 2>&1

  # determinar procesos a terminar
  get_tree_of_applications

  # verificar si ya existe una terminal
  [ -s ${APLOGT}.pid ] && report_status "i" "Wops, another ${SCRNAME} VirtualTerminal process exist!"
  [ -s ${APLOGT}.pid ] && log_action "DEBUG" "Wops, another ${SCRNAME} VirtualTerminal process exist!"
  [ -s ${APLOGT}.pid ] && exit 1

  # backup
  mkdir -p ${APTEMP}/${APDATE}
  mv ${APLOGP}.log ${APTEMP}/${APDATE}/${APPRCS}_${APHOUR}.log > /dev/null 2>&1
  log_action "DEBUG" "Backuping ${APLOGP}.log as ${APTEMP}/${APDATE}/${APPRCS}_${APHOUR}.log"

  #
  ${SCREEN} -d -m -S ${SCRNAME}
  ${SCREEN} -r ${SCRNAME} -p 0 -X log off
  ${SCREEN} -r ${SCRNAME} -p 0 -X logfile ${APLOGP}.log
  ${SCREEN} -r ${SCRNAME} -p 0 -X logfile flush ${APDLINES}
  ${SCREEN} -r ${SCRNAME} -p 0 -X log ${APDEBUG}
  log_action "DEBUG" "Creating VirtualTerminal ${SCRNAME}, logfile as ${APLOGP}.log in mode ${APDEBUG}"
  wait_for "Starting ${SCRNAME} virtual terminal " 2
  
  #
  ${SCREEN} -r ${SCRNAME} -p 0 -X stuff "$(printf '%b' "${APCOMMAND} || exit\015")"
  log_action "DEBUG" "Send the command [${APCOMMAND}] in ${SCRNAME}"
  wait_for "Starting process " 2

  # get the tree applications
  process_running
  [ -s ${APLOGT}.pid ] && report_status "*" "[${APLINK}] running in ${APPRCS} right now" ||
                          report_status "?" "[${APLINK}] in ${APPRCS} failed to initialize"

fi


# CHECK
if [ ${APACTION} = "CHECK" ]
then

  # determinar procesos a terminar
  get_tree_of_applications 

  #
  if [ -f ${APLOGT}.list ]
  then
    echo "Execution's tree"
    pos=" "
    for APID in $(cat ${APLOGT}.list)
    do
      PNAME=`awk -v pid=${APID} -v pos=${PSPOS} '{if ($(3+pos) ~ pid){print}}' ${APLOGT}.allps | sed -e "s/.*[0-9]:[0-9][0-9]//g" 2> /dev/null`
      [ ${#pos} -gt 0 ] && pos="${pos} "
      [ ${#pos} -eq 1 ] && index="${pos}+" || index="${pos}'-"
      printto "[${APID}  ] ${index} ${PNAME}"
    done
  fi

fi

 
# STOP
if [ ${APACTION} = "STOP" ]
then

  # determinar procesos a terminar
  get_tree_of_applications

  # verificar si ya existe una terminal
  [ ! -s ${APLOGT}.pid ] && report_status "i" "Hey, VirtualTerminal ${SCRNAME} doesnt exists..."
  [ ! -s ${APLOGT}.pid ] && log_action "DEBUG" "Hey, VirtualTerminal ${SCRNAME} doesnt exists..."
  [ ! -s ${APLOGT}.pid ] && exit 1

  ${SCREEN} -r ${SCRNAME} -p 0 -X log off
  ${SCREEN} -r ${SCRNAME} -p 0 -X stuff "$(printf '%b' "exit\015")"
  log_action "DEBUG" "Sending the exit command"
  wait_for "Stopping process " 2

  for APID in $(cat ${APLOGT}.list | sort -r)
  do
    kill -0 ${APID} > /dev/null 2>&1
    LASTSTATUS=$?
    PNAME=`awk -v pid=${APID} -v pos=${PSPOS} '{if ($(3+pos) ~ pid){print}}' ${APLOGT}.allps | sed -e "s/.*[0-9]:[0-9][0-9]//g" 2> /dev/null`
    [ $LASTSTATUS -eq 0 ] && $(kill -9 ${APID} > /dev/null 2>&1)
    [ $LASTSTATUS -eq 0 ] && log_action "DEBUG" "Process ${PNAME} finalized "
  done
  rm -f ${APLOGT}.*
  report_status "*" "VirtualTerminal process ${SCRNAME} and subprocesses finalized"

fi

# cualquier otro comando ...
# app.logsOn | app.logsOff | app.backUp | app.logsClear 
case ${APACTION}  in
  LOGSON)
    executeCmd "${APLOGSON}"
    exit 0
    ;;

  LOGSOFF)
    executeCmd "${APLOGSOFF}"
    exit 0
    ;;

  SYSLOGSON)
    executeCmd "${APSYSLOGSON}"
    exit 0
    ;;

  SYSLOGSOFF)
    executeCmd "${APSYSLOGSOFF}"
    exit 0
    ;;

  DBLOGSON)
    executeCmd "${APDBLOGSON}"
    exit 0
    ;;

  DBLOGSOFF)
    executeCmd "${APDBLOGSOFF}"
    exit 0
    ;;

  BACKUP)
    executeCmd "${APBACKUP}"
    exit 0
    ;;

  LOGSCLEAR)
    executeCmd "${APLOGSCLEAR}"
    exit 0
    ;;

  GETLEVEL)
    executeCmd "${APLEVEL}"
    tail -n100 ${APLOG}.log | grep "Level for this build" | tail -n1
    exit 0
    ;;

  VERSION)
    #
    printto "${APNAME} ${VERSION} (${RELEASE})"
    printto "${APPROF}\n"

    if [ "${TTYTYPE}" = "CONSOLE" ]
    then 
      printto  "Developed by"
      printto  "Andres Aquino <aquino@hp.com>"
    fi  
    exit 0
    ;;

esac

#
